### DISCUSSION TITLE
Units of Measure Confusion

### DISCUSSION THREAD

**Discussion by NCJazzMan65** [IDAES/idaes-pse#538] [General]
As a fairly new user, I'm still struggling with understanding the units of measure for calculated quantities such as work and heat.  As an example, say I have a Pressure Changer (compressor in this case) on my flowsheet and my property package defines the following units: time=pyunits.s, length=pyunits.m, mass=pyunits.kg, amount=pyunits.mol (I assume this is g-mol), and temperature=pyunits.K.  Further, say my property package is using equations from the NIST webbook.  When I run the flowsheet, my Pressure Changer reports a calculated amount of work.  Based on the magnitude of the work number in relation to my flow conditions, I'm guessing the units are Watts.  But how do I know for sure?  I tried figuring it out from the base units I set and the reported units for the NIST properties (enthalpy, specific heat, etc.) but I'm struggling with it.  How do I figure this out?

I know the underlying PYOMO framework is unit agnostic but, surely, something better can be done to represent and report units on values in IDAES; especially on calculated values like work and heat.

**Comment by andrewlee94** [comment]
@NCJazzMan65 At the moment, you can use the Pyomo `get_units()` function to retrieve the units of measurement for any quantity. We are also working with Pyomo to improve the `value()` function to return both the magnitude and units of a quantity which we hope to have in the next release.

**Reply by NCJazzMan65** [reply]
Andrew...thank you for your reply.  Can you give me an example of how you would use the get_units() function in my situation?  Assume my Pressure Changer is called "model.fs.blower" and I want to know the units of the calculated work term.

**Reply by andrewlee94** [reply]
The following code should get (and then print) the units of the work term:

```
import pyomo.environ as pyo

units = pyo.units.get_units(model.fs.blower.work_mechanical[0])

print(units)
```

A couple of things to note:

1. This code will work for any variable or expression.
2. When dealing with indexed variables or expressions, `get_units` should only be used on a single element of the indexed component (i.e. only the individual values have units, not the entire indexed component). This is why the time index (`[0]`) appears in the code above.
3. The units are returned as a Pyomo UnitsContainer object, and can then be used as part of unit conversion methods if required (e.g. `pyo.units.convert_to(42, from_units=pyo.units.MW, to_units=units)`

**Reply by NCJazzMan65** [reply]
Thank you for the example.  This is very helpful.

One suggestion for a future IDAES release...Incorporate a call to the get_units() function into the standard report functions for streams and blocks so units can be shown along with the values.  This would seem a pretty straightforward thing to implement.
